common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -138,11 +138,13 @@ public Collation(
     collationTable[2] = new Collation(
       "UNICODE", Collator.getInstance(ULocale.ROOT), "153.120.0.0", true);
     collationTable[2].collator.setStrength(Collator.TERTIARY);
+    collationTable[2].collator.freeze();
```

Comment:
This is interesting, I thought `Collator` is just holding a bunch of functions, but it has mutable states?

Reply:
@cloud-fan collator uses a buffer while writing collation keys, freezing it makes this operation safe by using a reentrant lock around it ([source](https://github.com/unicode-org/icu/blob/main/icu4j/main/collate/src/main/java/com/ibm/icu/text/RuleBasedCollator.java#L1831))

this of course raises performance issues which we should probably discuss, because now only one thread can generate sort keys

Reply:
Yeah, as soon as we get benchmarks working we should revisit this decision.
One option that we also prototyped is to keep `Collator` in `ThreadLocal` fields, which also solved the problem. But `freeze` is a bit cleaner and we don't have microbenchmarks yet so we can't make data driven decision at this point.
LGTM.